Fails to work with HSM (RSA OAEP)

**Describe the bug**

Fails to either encrypt or decrypt.

**To Reproduce**

Encryption (with provider):
```
$ env | grep MODULE
PKCS11_PROVIDER_MODULE=/usr/local/lib/yubihsm_pkcs11.dylib
YUBIHSM_PKCS11_MODULE=/usr/local/lib/pkcs11/yubihsm_pkcs11.dylib
$ echo $MODULE
/usr/local/lib/yubihsm_pkcs11.dylib
$ cmp $MODULE $YUBIHSM_PKCS11_MODULE 
$ 
$ openssl pkeyutl  -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.bin -out /tmp/key.enc
Could not read public key from pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public
pkeyutl: Error initializing context
$
```
Encryption failed - needed to prompt for the PIN, but it didn't.

For decryption (with provider):
```
$ openssl pkeyutl  -decrypt -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=private" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.enc -out /tmp/key.dec
Enter pass phrase for PKCS#11 Token (Slot 0 - YubiHSM Connector 0.0.0.0):
Segmentation fault: 11
$ ll /tmp/key.dec
-rw-r--r--  1 ur20980  wheel  32 Feb 14 10:23 /tmp/key.dec
$ cmp /tmp/key.bin /tmp/key.dec
$
```
Appears to have decrypted correctly before crashing OpenSSL. Observed crash:
```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [15524]

VM Region Info: 0 is not in any region.  Bytes before following region: 4307480576
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      100bef000-100c6f000    [  512K] r-x/r-x SM=COW  ...xec/*/openssl

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_pthread.dylib       	    0x7ff813668f63 pthread_rwlock_rdlock + 0
1   libcrypto.3.dylib             	       0x1012e434f CRYPTO_THREAD_read_lock + 9
2   libcrypto.3.dylib             	       0x10130662b ossl_method_store_cache_get + 59
3   libcrypto.3.dylib             	       0x1012b8aa1 inner_evp_generic_fetch + 366
4   libcrypto.3.dylib             	       0x1012b8912 evp_generic_fetch + 65
5   libcrypto.3.dylib             	       0x1012b6307 evp_cipher_init_internal + 1389
6   libcrypto.3.dylib             	       0x1012b709d EVP_EncryptInit_ex + 27
7   libyubihsm.2.4.0.dylib        	       0x100e4eeba aes_encrypt_ex + 58
8   libyubihsm.2.4.0.dylib        	       0x100e508c2 send_encrypted_msg + 1218
9   libyubihsm.2.4.0.dylib        	       0x100e50259 yh_send_secure_msg + 89
10  libyubihsm.2.4.0.dylib        	       0x100e590ab yh_util_close_session + 107
11  yubihsm_pkcs11.dylib          	       0x100f29264 C_CloseSession + 372
12  pkcs11.dylib                  	       0x100e8ee77 p11prov_CloseSession + 87
13  pkcs11.dylib                  	       0x100e9b242 session_free + 194
14  pkcs11.dylib                  	       0x100e9934e p11prov_free_slots + 366
15  pkcs11.dylib                  	       0x100e9796b p11prov_ctx_free + 187
16  libcrypto.3.dylib             	       0x1012e15ff ossl_provider_free + 61
17  libcrypto.3.dylib             	       0x1013466ac OPENSSL_sk_pop_free + 45
18  libcrypto.3.dylib             	       0x1012e2e02 provider_store_free + 63
19  libcrypto.3.dylib             	       0x1012d91d6 CRYPTO_free_ex_data + 326
20  libcrypto.3.dylib             	       0x1012d66d6 context_deinit + 98
21  libcrypto.3.dylib             	       0x1012d6667 ossl_lib_ctx_default_deinit + 16
22  libcrypto.3.dylib             	       0x1012d95ca OPENSSL_cleanup + 200
23  libsystem_c.dylib             	    0x7ff813562c26 __cxa_finalize_ranges + 416
24  libsystem_c.dylib             	    0x7ff813562a39 exit + 35
25  openssl                       	       0x100c131bb main + 594
26  dyld                          	    0x7ff813339310 start + 2432
```


**Expected behavior**

For encryption (showing with the engine):
```
$ openssl pkeyutl -engine pkcs11 -keyform engine -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.bin -out /tmp/key.enc
Engine "pkcs11" set.
Enter PKCS#11 token PIN for YubiHSM:
$
$ ll /tmp/key.enc
-rw-r--r--  1 ur20980  wheel  384 Feb 14 10:21 /tmp/key.enc
$ 
```

For decryption (showing with the engine):
```
$ openssl pkeyutl -engine pkcs11 -keyform engine -decrypt -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=private" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.enc -out /tmp/key.dec
Engine "pkcs11" set.
Enter PKCS#11 token PIN for YubiHSM:
$ cmp /tmp/key.bin /tmp/key.dec
$ 
```

**Operating environment (please complete the following information):**
 - OS: MacOS Ventura
 - Version: 13.2.1

**Token and application used (please complete the following information):**
 - Device: YubiHSM
 - PKCS11 Driver version: YubiHSM-2.4.0
 - Application: OpenSSL
 - Version: 3.0.8

**Additional context**

Works with engine, partially works for decryption with HSM when invoked manually. Adding or removing `object=RSA_OAEP;` to the key URI seems to have no effect, probably better to remove it.

```
$ yhsm2-rsa-encr-prov 
YUBIHSM_PKCS11_CONF = "/Users/ur20980/yubihsm_pkcs11.conf"
YUBIHSM_PKCS11_MODULE = "/usr/local/lib/pkcs11/yubihsm_pkcs11.dylib"


Testing pkcs11 engine (libp11) with YubiHSM

Generate a random data file (1000 bytes), Base64-encoded...
openssl rand -engine rdrand -base64 -out /tmp/derive.1367.text 1000
Engine "rdrand" set.

Engine "rdrand" set.
Engine "rdrand" set.
Generated
  key: 47d3b66087235de5dd0680bba9a64a947e7589aa31b4f8d4eed73201cb4347ba
  iv:  3fbc1b3c023109e4e077e36e2d7baca3

Encrypt file /tmp/derive.1367.text with this key and IV...
openssl enc -aes-256-cfb -a -e -in /tmp/derive.1367.text -out /tmp/derive.1367.text.enc -K 47d3b66087235de5dd0680bba9a64a947e7589aa31b4f8d4eed73201cb4347ba -iv 3fbc1b3c023109e4e077e36e2d7baca3

Encrypting this random symmetric key to token RSA KEY MAN key...
echo 47d3b66087235de5dd0680bba9a64a947e7589aa31b4f8d4eed73201cb4347ba | xxd -r -p -c 200 | openssl pkeyutl -engine pkcs11 -keyform engine -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -out /tmp/derive.1367.key.enc
Engine "pkcs11" set.
Enter PKCS#11 token PIN for YubiHSM:

encrypted key:
4a3d7d5870e683fc580814117507a2fb09eba5e34060d57dc00e9ace540a8c424638f1ab9633e21ec39349e39588efb8857ef0defb36cffb16eebc619278de4bdbd64e3d075d39dee1c137256df1d779e70d365ccb16fab604d73f0745a60f15a6c95e9a2e7dc872d0a61b088bd9eb2a2ee9120c603e5c55f66ec9f02e3f45a01706be69131e0b3deb3cd823ed46a79633dc06d11eee139b56aad37672144431f7109ad13f0c0b09d8a22478e58db3e850199bbe2751f0a0e2824094b1ae6df63c07417e6846e7e7
5e54632acb1ae0007a0af1a75d1e130cf2246eae3d255db903862422a76f844e7962412f34d9c04eb8b4a9bbfb64c50004fa0b7697b3f1360fe21329d14f45996ab18f7f03e9025d53f8242813a7376d1f29dacbd7af31fc2eb534a261023a110d7d1387e4415f986e6b683453cb11ebaa4d414397a517ba03673d3cf1e5804d73e9822afb41a03034fa7f051b9141be563424c74e181228a9c2b2695d76160b4e2e49ad42909be7527883c0e43a504da7786d211a865c16

Decrypting the symmetric key on the token...
openssl pkeyutl -engine pkcs11 -keyform engine -decrypt -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;type=private" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/derive.1367.key.enc | xxd -p -c 200
Engine "pkcs11" set.
Enter PKCS#11 token PIN for YubiHSM:

Decrypt file /tmp/derive.1367.text.enc with this key and IV...
openssl enc -aes-256-cfb -a -d -in /tmp/derive.1367.text.enc -out /tmp/derive.1367.text.dec -K 47d3b66087235de5dd0680bba9a64a947e7589aa31b4f8d4eed73201cb4347ba -iv 3fbc1b3c023109e4e077e36e2d7baca3

KEY1="47d3b66087235de5dd0680bba9a64a947e7589aa31b4f8d4eed73201cb4347ba"
KEY2="47d3b66087235de5dd0680bba9a64a947e7589aa31b4f8d4eed73201cb4347ba"

Original and decrypted keys  match

Original and decrypted files match


Cleaning up...


$ 
```

> Please check if #197 resolves your issue, I will pile a little bit more documentation i that PR, but the first patch is what you need

Alas, no:
```
$ openssl pkeyutl -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.bin -out /tmp/key.enc1
Enter pass phrase for PKCS#11 Token (Slot 0 - YubiHSM Connector 0.0.0.0):
Public Key operation error
8096FA56F87F0000:error:42000063:pkcs11:p11prov_EncryptInit:The specified key is not the correct type of key to use with the specified mechanism:./interface.gen.c:413:Error returned by C_EncryptInit
Segmentation fault: 11
```

[yhsm2-rsa-encr-spy2.txt](https://github.com/latchset/pkcs11-provider/files/10735018/yhsm2-rsa-encr-spy2.txt)

--- a/src/provider.c
+++ b/src/provider.c
@@ -28,6 +28,7 @@ struct p11prov_ctx {
     const char *init_args;
     char *pin;
     int allow_export;
+    int login_behavior;
     /* TODO: ui_method */
     /* TODO: fork id */
 
@@ -328,6 +329,12 @@ int p11prov_ctx_allow_export(P11PROV_CTX *ctx)
 {
     P11PROV_debug("allow_export = %d", ctx->allow_export);
     return ctx->allow_export;
+}
+
+int p11prov_ctx_login_behavior(P11PROV_CTX *ctx)
+{
+    P11PROV_debug("login_behavior = %d", ctx->login_behavior);
+    return ctx->login_behavior;
 }
 
 static void p11prov_teardown(void *ctx)
@@ -1049,9 +1056,10 @@ static int p11prov_module_init(P11PROV_CTX *ctx)
 int OSSL_provider_init(const OSSL_CORE_HANDLE *handle, const OSSL_DISPATCH *in,
                        const OSSL_DISPATCH **out, void **provctx)
 {
-    OSSL_PARAM core_params[5] = { 0 };
+    OSSL_PARAM core_params[6] = { 0 };
     const char *module = NULL;
     char *allow_export = NULL;
+    char *login_behavior = NULL;
     char *pin = NULL;
     P11PROV_CTX *ctx;
     int ret;
@@ -1083,7 +1091,10 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle, const OSSL_DISPATCH *in,
     core_params[3] =
         OSSL_PARAM_construct_utf8_ptr(P11PROV_PKCS11_MODULE_ALLOW_EXPORT,
                                       &allow_export, sizeof(allow_export));
-    core_params[4] = OSSL_PARAM_construct_end();
+    core_params[4] =
+        OSSL_PARAM_construct_utf8_ptr(P11PROV_PKCS11_MODULE_LOGIN_BEHAVIOR,
+                                      &login_behavior, sizeof(login_behavior));
+    core_params[5] = OSSL_PARAM_construct_end();
     ret = core_get_params(handle, core_params);
     if (ret != RET_OSSL_OK) {
         ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
@@ -1123,6 +1134,21 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle, const OSSL_DISPATCH *in,
         if (errno != 0 || *end != '\0') {
             P11PROV_raise(ctx, CKR_GENERAL_ERROR, "Invalid value for %s: (%s)",
                           P11PROV_PKCS11_MODULE_ALLOW_EXPORT, allow_export);
+            p11prov_ctx_free(ctx);
+            return RET_OSSL_ERR;
+        }
+    }
+
+    if (login_behavior != NULL) {
+        if (strcmp(login_behavior, "auto") == 0) {
+            ctx->login_behavior = PUBKEY_LOGIN_AUTO;
+        } else if (strcmp(login_behavior, "always") == 0) {
+            ctx->login_behavior = PUBKEY_LOGIN_ALWAYS;
+        } else if (strcmp(login_behavior, "never") == 0) {
+            ctx->login_behavior = PUBKEY_LOGIN_NEVER;
+        } else {
+            P11PROV_raise(ctx, CKR_GENERAL_ERROR, "Invalid value for %s: (%s)",
+                          P11PROV_PKCS11_MODULE_LOGIN_BEHAVIOR, login_behavior);
             p11prov_ctx_free(ctx);
             return RET_OSSL_ERR;
         }

--- a/src/provider.h
+++ b/src/provider.h
@@ -34,6 +34,7 @@
 #define P11PROV_PKCS11_MODULE_INIT_ARGS "pkcs11-module-init-args"
 #define P11PROV_PKCS11_MODULE_TOKEN_PIN "pkcs11-module-token-pin"
 #define P11PROV_PKCS11_MODULE_ALLOW_EXPORT "pkcs11-module-allow-export"
+#define P11PROV_PKCS11_MODULE_LOGIN_BEHAVIOR "pkcs11-module-login-behavior"
 
 #define P11PROV_DEFAULT_PROPERTIES "provider=pkcs11"
 #define P11PROV_NAMES_RSA "RSA:rsaEncryption:1.2.840.113549.1.1.1"
@@ -77,6 +78,11 @@ CK_RV p11prov_ctx_set_quirk(P11PROV_CTX *ctx, CK_SLOT_ID id, const char *name,
 #define ALLOW_EXPORT_PUBLIC 0
 #define DISALLOW_EXPORT_PUBLIC 1
 int p11prov_ctx_allow_export(P11PROV_CTX *ctx);
+
+#define PUBKEY_LOGIN_AUTO 0
+#define PUBKEY_LOGIN_ALWAYS 1
+#define PUBKEY_LOGIN_NEVER 2
+int p11prov_ctx_login_behavior(P11PROV_CTX *ctx);
 
 #include "debug.h"
 

--- a/src/store.c
+++ b/src/store.c
@@ -85,13 +85,18 @@ static void store_fetch(struct p11prov_store_ctx *ctx,
 {
     CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
     CK_SLOT_ID nextid = CK_UNAVAILABLE_INFORMATION;
+    int login_behavior;
     bool login = false;
     CK_RV ret;
 
-    if (ctx->expect == 0 || ctx->expect == OSSL_STORE_INFO_PKEY) {
+    login_behavior = p11prov_ctx_login_behavior(ctx->provctx);
+
+    if (ctx->expect == 0 || ctx->expect == OSSL_STORE_INFO_PKEY
+        || login_behavior == PUBKEY_LOGIN_ALWAYS) {
         login = true;
     }
 
+again:
     /* cycle through all available slots,
      * only stack errors, but not block on any of them */
     do {
@@ -129,6 +134,17 @@ static void store_fetch(struct p11prov_store_ctx *ctx,
         slotid = nextid;
 
     } while (nextid != CK_UNAVAILABLE_INFORMATION);
+
+    /* Given the variety of tokens, if we found no object at all, and we did
+     * *not* set login required, we retry again, after setting login required.
+     * This accounts for HW that requires a login even for public objects */
+    if (login == false && ctx->num_objs == 0
+        && login_behavior != PUBKEY_LOGIN_NEVER) {
+        slotid = CK_UNAVAILABLE_INFORMATION;
+        ctx->loaded = 0;
+        login = true;
+        goto again;
+    }
 }
 
 DISPATCH_STORE_FN(open);

